### ISSUE TITLE
Question about `first_times_fp` parameter behavior and optimal setting in Wan2.1

### ISSUE THREAD

**Issue by t0saki** [svg-project/Sparse-VideoGen#35]
Thank you very much for your impressive work and for sharing both the paper and code—this is a very interesting project!

I have a question regarding the behavior and optimal setting of the `first_times_fp` parameter. According to my understanding, for a value like 0.09, I would expect full attention computation in the first 4 (or 5) denoising steps. However, based on my timing tests, it seems that the full attention mode actually lasts up to the first 18 steps. Is this an expected behavior, or could there be a misunderstanding about how `first_times_fp` is applied?

Additionally, could you share how you determined the recommended values for `first_times_fp` and `first_layers_fp`? For example, in the provided example for WAN, the default `first_times_fp` parameter is 0.09, which appears inconsistent with the recommendations mentioned in the code comments. I am very curious about the specific effects of these hyperparameters, and especially how you determined the optimal values for `first_times_fp` and `first_layers_fp` under different `sparsity`.

This issue also appears to be mentioned in [#33](https://github.com/svg-project/Sparse-VideoGen/issues/33).

Here is a command and console output for reference:
```bash
python wan_t2v_inference.py --prompt "a walking figure made out of water tours an art gallery with many beautiful works of art in different styles" --height 720 --width 1280 --pattern SVG --num_sampled_rows 64 --sparsity 0.35 --first_times_fp 0.09 --first_layers_fp 0.025 --output_file result/Sparsity_0.35_FTFP_0.09/Wan_T2V/SVG/t2v_item_4_svg.mp4

...

  0%|          | 0/50 [00:00<?, ?it/s]
  2%|▏         | 1/50 [02:17<1:52:34, 137.84s/it]
  4%|▍         | 2/50 [04:35<1:50:04, 137.60s/it]
  6%|▌         | 3/50 [06:52<1:47:44, 137.54s/it]
  8%|▊         | 4/50 [09:09<1:45:17, 137.34s/it]
 10%|█         | 5/50 [11:26<1:42:44, 136.99s/it]
 12%|█▏        | 6/50 [13:42<1:40:18, 136.78s/it]
 14%|█▍        | 7/50 [15:58<1:37:55, 136.65s/it]
 16%|█▌        | 8/50 [18:15<1:35:35, 136.56s/it]
 18%|█▊        | 9/50 [20:31<1:33:16, 136.51s/it]
 20%|██        | 10/50 [22:48<1:30:58, 136.47s/it]
 22%|██▏       | 11/50 [25:04<1:28:41, 136.44s/it]
 24%|██▍       | 12/50 [27:20<1:26:23, 136.42s/it]
 26%|██▌       | 13/50 [29:37<1:24:07, 136.41s/it]
 28%|██▊       | 14/50 [31:53<1:21:50, 136.40s/it]
 30%|███       | 15/50 [34:09<1:19:33, 136.39s/it]
 32%|███▏      | 16/50 [36:26<1:17:17, 136.39s/it]
 34%|███▍      | 17/50 [38:42<1:15:00, 136.38s/it]
 36%|███▌      | 18/50 [39:50<1:01:45, 115.79s/it]
 38%|███▊      | 19/50 [40:57<52:19, 101.27s/it]  
 40%|████      | 20/50 [42:05<45:33, 91.12s/it] 
 42%|████▏     | 21/50 [43:12<40:36, 84.01s/it]
 44%|████▍     | 22/50 [44:20<36:53, 79.04s/it]
 46%|████▌     | 23/50 [45:27<34:00, 75.56s/it]
 48%|████▊     | 24/50 [46:35<31:41, 73.13s/it]
 50%|█████     | 25/50 [47:42<29:45, 71.42s/it]
 52%|█████▏    | 26/50 [48:50<28:05, 70.23s/it]
 54%|█████▍    | 27/50 [49:57<26:36, 69.40s/it]
 56%|█████▌    | 28/50 [51:05<25:13, 68.81s/it]
 58%|█████▊    | 29/50 [52:12<23:56, 68.40s/it]
 60%|██████    | 30/50 [53:19<22:42, 68.12s/it]
 62%|██████▏   | 31/50 [54:27<21:30, 67.92s/it]
 64%|██████▍   | 32/50 [55:34<20:19, 67.78s/it]
 66%|██████▌   | 33/50 [56:42<19:10, 67.68s/it]
 68%|██████▊   | 34/50 [57:49<18:01, 67.61s/it]
 70%|███████   | 35/50 [58:57<16:53, 67.56s/it]
 72%|███████▏  | 36/50 [1:00:04<15:45, 67.53s/it]
 74%|███████▍  | 37/50 [1:01:12<14:37, 67.50s/it]
 76%|███████▌  | 38/50 [1:02:19<13:29, 67.49s/it]
 78%|███████▊  | 39/50 [1:03:26<12:22, 67.48s/it]
 80%|████████  | 40/50 [1:04:34<11:14, 67.47s/it]
 82%|████████▏ | 41/50 [1:05:41<10:07, 67.46s/it]
 84%|████████▍ | 42/50 [1:06:49<08:59, 67.46s/it]
 86%|████████▌ | 43/50 [1:07:56<07:52, 67.46s/it]
 88%|████████▊ | 44/50 [1:09:04<06:44, 67.45s/it]
 90%|█████████ | 45/50 [1:10:11<05:37, 67.45s/it]
 92%|█████████▏| 46/50 [1:11:19<04:29, 67.45s/it]
 94%|█████████▍| 47/50 [1:12:26<03:22, 67.45s/it]
 96%|█████████▌| 48/50 [1:13:33<02:14, 67.45s/it]
 98%|█████████▊| 49/50 [1:14:41<01:07, 67.45s/it]
100%|██████████| 50/50 [1:15:48<00:00, 67.45s/it]
100%|██████████| 50/50 [1:15:48<00:00, 90.98s/it]
```

**Comment by haochengxi** [comment]
Thanks for the question. The first question is related to the scheduler in denoising steps. From the code below you can find that 0.09 means 35% of the denoising steps use full attention calculation, which is about 17-18 denoising steps.

https://github.com/svg-project/Sparse-VideoGen/blob/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e/scripts/wan_t2v_inference.sh#L3

For the second question, increasing `first_times_fp` improves PSNR and overall video quality, though at the expense of higher latency. The optimal values are determined experimentally, based on when the video quality meets our standards—which can vary from person to person.

**Comment by haochengxi** [comment]
Meanwhile, `first_layers_fp` is always set to satisfy that only the first layer uses dense attention. That's because we find that the first layer's attention map behaves differently compared with other layers.

**Comment by t0saki** [comment]
Thank you for your detailed explanation! I realize now that I previously misunderstood the comment about `first_times_fp`.

I'm just a bit curious—why the actual proportion of inference steps using full attention doesn't exactly match the `first_times_fp` value? I looked into the conversion between inference steps and timesteps, and it seems that timesteps are sampled linearly. According to
https://github.com/svg-project/Sparse-VideoGen/blob/079364dc2e4ca6cd0c26c8c45eafeb9fbf51ef8e/svg/models/wan/attention.py#L214
It appears that `first_times_fp` should directly control the proportion of inference steps with full attention. However, my results seem a bit inconsistent with this. Is there something I might be overlooking here?

Thanks again for your help!

**Comment by Popping-Lock** [comment]
> Meanwhile, `first_layers_fp` is always set to satisfy that only the first layer uses dense attention. That's because we find that the first layer's attention map behaves differently compared with other layers.

 ‘first layer’s attention map behaves differently compared with other layers’ — does the attention map here refer to a heatmap? Can I choose corresponding parameters based on the differences shown in the heatmap?”

**Comment by haochengxi** [comment]
For the second question, yes, attention map refers to a heat map after softmax operation. I suggest to skip this layer, since it usually does not exhibit a clear spatial or temporal pattern, and is different from all other layers. Other layers' heatmap (even at layer 1) behaves relatively similarly.
